### ISSUE TITLE
Improve assignment of built-ins default values

### ISSUE THREAD

**Issue by shay2025** [reviewpad/reviewpad#445]
**Is your feature request related to a problem? Please describe.**
At the moment, we use the transforms functions (in `engine/transform.go`) to provide default values to the built-ins that have no parameters provided and require parameters. To do so, we use regex.

However, this approach seems brittle. 

**Describe the solution you'd like**
We should conduct an investigation to see how other parsers try to provide default values to functions and figure out how these approaches fit in our parser.

**Comment by zolamk** [comment]
I have been looking into this issue for the past couple of days, even though I couldn't find feasible examples of how this is handled in other parsers, I did have a lot of programming languages that support default parameters in function definitions to look at for example, like
- [Ruby](https://docs.ruby-lang.org/en/2.4.0/syntax/calling_methods_rdoc.html#label-Default+Positional+Arguments)
- [Python](https://www.programiz.com/python-programming/function-argument)
- [C++](https://www.programiz.com/cpp-programming/default-argument)
- [Scala](https://docs.scala-lang.org/tour/default-parameter-values.html)
- ...

I think the common pattern is that default arguments are provided alongside function parameter definitions using some sort of syntax like `function ([variable_name] datatype, [variable_name] datatype = default)` I believe we can also apply the same concept to ReviewPad built-ins, depending on what we want to support we would need to modify the ReviewPad types and aladino parser as well

- Since we introduced support for function types in this PR https://github.com/reviewpad/reviewpad/pull/413, we might want to make it possible for user defined functions to have a default by modifying the parser to support syntax like `Func (String, Int = 5, Bool = true)`
- Since we are specifying ReviewPad built-ins using Go code, I propose we add a way to store default values in the current types we have(`String`, `Int`, `Bool`, `ArrayOfType`, `ArrayType` and `FunctionType`)
  maybe like
  ```
  type StringType struct {
      def Expr
  }

  type IntType struct {
	  def Expr
  }

  type BoolType struct {
	  def Expr
  }

  func BuildStringType(d *StringConst) *StringType { return &StringType{d} }
  func BuildIntType(d *IntConst) *IntType          { return &IntType{d} }
  func BuildBoolType(d *BoolConst) *BoolType       { return &BoolType{d} }
  ```
  and then we can build our functions like

  > Note: we can use `nil` like `aladino.BuildStringType(nil)` to indicate that a certain parameter is required and doesn't have a default value

  ```
  func IssueCountBy() *aladino.BuiltInFunction {
	  defaultUser := aladino.BuildStringConst("")
	  defaultState := aladino.BuildStringConst("all")
	  return &aladino.BuiltInFunction{
		  Type: aladino.BuildFunctionType(
			  []aladino.Type{
				  aladino.BuildStringType(defaultUser),
				  aladino.BuildStringType(defaultState),
			  },
			  aladino.BuildIntType(nil),
		  ),
		  Code:           issueCountByCode,
		  SupportedKinds: []handler.TargetEntityKind{handler.PullRequest, handler.Issue},
	  }
  }
  ```
  After which, during the execution of the program, we can merge the user provided arguments with the default arguments in places where the user has omitted some arguments, so the user can call the built-in function like

  ```
   $issueCountBy()
   $issueCountBy($author())
   $issueCountBy($author, "all")
  ```

  without us needing to perform regular expression gymnastics to provide the default values to functions

## References
- https://github.com/AthrunArthur/cxxparser
- [https://www.programiz.com/cpp-programming/default-argument](https://www.programiz.com/cpp-programming/default-argument)
- [https://www.geeksforgeeks.org/default-arguments-in-python/](https://www.geeksforgeeks.org/default-arguments-in-python/)
- [https://stackoverflow.com/questions/3885286/which-programming-languages-allow-default-values-for-method-parameters](https://stackoverflow.com/questions/3885286/which-programming-languages-allow-default-values-for-method-parameters)
- [https://github.com/python/cpython/blob/main/Grammar/python.gram](https://github.com/python/cpython/blob/main/Grammar/python.gram)
- [https://github.com/python/cpython/blob/73943cbc4c5e97f71b76150c549d07e8ed00066b/Grammar/python.gram#L291](https://github.com/python/cpython/blob/73943cbc4c5e97f71b76150c549d07e8ed00066b/Grammar/python.gram#L291)
- [https://github.com/python/cpython/blob/aa0f056a00c4bcaef83d729e042359ddae903382/Parser/action_helpers.c#L681](https://github.com/python/cpython/blob/aa0f056a00c4bcaef83d729e042359ddae903382/Parser/action_helpers.c#L681)

**Comment by shay2025** [comment]
Another possibility could be having another field called `Default` where we would put the default values.
```
func IssueCountBy() *aladino.BuiltInFunction {
    return &aladino.BuiltInFunction{
        Type:           aladino.BuildFunctionType([]aladino.Type{aladino.BuildStringType(), aladino.BuildStringType()}, aladino.BuildIntType(), aladino.BuildBoolType()),
        Default:        []aladino.Type{aladino.BuildStringValue(""), aladino.BuildStringValue("all")}, aladino.BuildIntValue(nil)),
        Code:           issueCountByCode,
        SupportedKinds: []handler.TargetEntityKind{handler.PullRequest, handler.Issue},
    }
}
```

**Comment by zolamk** [comment]
> Another possibility could be having another field called `Default` where we would put the default values.
> 
> ```
> func IssueCountBy() *aladino.BuiltInFunction {
>     return &aladino.BuiltInFunction{
>         Type: aladino.BuildFunctionType(
> 		[]aladino.Type{
> 			aladino.BuildStringType(), 
> 			aladino.BuildStringType(),
> 		}, 
> 		aladino.BuildIntType(),
> 	),
>         Default: []aladino.Type{
> 		aladino.BuildStringValue(""), 
> 		aladino.BuildStringValue("all"),
> 	},
>         Code: issueCountByCode,
>         SupportedKinds: []handler.TargetEntityKind{handler.PullRequest, handler.Issue},
>     }
> }
> ```

Yeah, This might be better actually because It won't involve touching the types and need a lot of changes

**Comment by ferreiratiago** [comment]
We have decided to leave it as it is and come back to this later one.

**Comment by zolamk** [comment]
@marcelosousa @ferreiratiago @shay2025 Any new feedback or suggestions on this issue

**Comment by shay2025** [comment]
> Another possibility could be having another field called `Default` where we would put the default values.
> 
> ```
> func IssueCountBy() *aladino.BuiltInFunction {
>     return &aladino.BuiltInFunction{
>         Type: aladino.BuildFunctionType(
> 		[]aladino.Type{
> 			aladino.BuildStringType(), 
> 			aladino.BuildStringType(),
> 		}, 
> 		aladino.BuildIntType(),
> 	),
>         Default: []aladino.Type{
> 		aladino.BuildStringValue(""), 
> 		aladino.BuildStringValue("all"),
> 	},
>         Code: issueCountByCode,
>         SupportedKinds: []handler.TargetEntityKind{handler.PullRequest, handler.Issue},
>     }
> }
> ```

@zolamk I have already given this suggestion 🙂
